File: .tmp_py4013/batch_04_sources/2467_most_profitable_path_in_a_tree.js
// LeetCode 2467 - Most Profitable Path in a Tree
// https://leetcode.com/problems/most-profitable-path-in-a-tree/

/**
 * @param {number[][]} edges
 * @param {number} bob
 * @param {number[]} amount
 * @return {number}
 */
var mostProfitablePath = function(edges, bob, amount) {
    const n = amount.length;
    const g = Array.from({ length: n }, () => []);
    for (const [a, b] of edges) {
        g[a].push(b);
        g[b].push(a);
    }
    const bobTime = Array(n).fill(n);
    const findBob = (u, p, t) => {
        if (u === 0) {
            bobTime[u] = t;
            return true;
        }
        for (const v of g[u]) {
            if (v === p) continue;
            if (findBob(v, u, t + 1)) {
                bobTime[u] = t;
                return true;
            }
        }
        return false;
    };
    findBob(bob, -1, 0);
    let ans = -Infinity;
    const dfs = (u, p, t, income) => {
        let cur = amount[u];
        if (t > bobTime[u]) cur = 0;
        else if (t === bobTime[u]) cur = Math.floor(cur / 2);
        income += cur;
        let isLeaf = true;
        for (const v of g[u]) {
            if (v !== p) {
                isLeaf = false;
                dfs(v, u, t + 1, income);
            }
        }
        if (isLeaf && income > ans) ans = income;
    };
    dfs(0, -1, 0, 0);
    return ans;
};

File: .tmp_rs4013/rust_check_batch08/2467_most_profitable_path_in_a_tree.rs
struct Solution;
// LeetCode 2467 - Most Profitable Path in a Tree
// https://leetcode.com/problems/most-profitable-path-in-a-tree/

impl Solution {
    pub fn most_profitable_path(edges: Vec<Vec<i32>>, bob: i32, amount: Vec<i32>) -> i32 {
        let n = amount.len();
        let mut g = vec![Vec::new(); n];
        for e in &edges {
            let (a, b) = (e[0] as usize, e[1] as usize);
            g[a].push(b);
            g[b].push(a);
        }
        let mut bob_time = vec![n as i32; n];
        fn find_bob(
            u: usize,
            p: i32,
            t: i32,
            g: &[Vec<usize>],
            bob_time: &mut [i32],
        ) -> bool {
            if u == 0 {
                bob_time[u] = t;
                return true;
            }
            for &v in &g[u] {
                if v as i32 == p {
                    continue;
                }
                if find_bob(v, u as i32, t + 1, g, bob_time) {
                    bob_time[u] = t;
                    return true;
                }
            }
            false
        }
        find_bob(bob as usize, -1, 0, &g, &mut bob_time);
        let mut ans = i32::MIN;
        fn dfs(
            u: usize,
            p: i32,
            t: i32,
            income: i32,
            g: &[Vec<usize>],
            amount: &[i32],
            bob_time: &[i32],
            ans: &mut i32,
        ) {
            let mut cur = amount[u];
            if t > bob_time[u] {
                cur = 0;
            } else if t == bob_time[u] {
                cur /= 2;
            }
            let income = income + cur;
            let mut is_leaf = true;
            for &v in &g[u] {
                if v as i32 != p {
                    is_leaf = false;
                    dfs(v, u as i32, t + 1, income, g, amount, bob_time, ans);
                }
            }
            if is_leaf && income > *ans {
                *ans = income;
            }
        }
        dfs(0, -1, 0, 0, &g, &amount, &bob_time, &mut ans);
        ans
    }
}

fn main() {}
